src/main/java/org/sonatype/nexus/ci/nxrm/v3/CreateTagBuilder.java

```diff
@@ -0,0 +1,229 @@
+    this.tagAttributesPath = fixEmptyAndTrim(tagAttributesPath);
+  }
+
+  @CheckForNull
+  public String getTagAttributesJson() {
+    return tagAttributesJson;
+  }
+
+  @DataBoundSetter
+  public void setTagAttributesJson(@CheckForNull final String tagAttributesJson) {
+    this.tagAttributesJson = fixEmptyAndTrim(tagAttributesJson);
+  }
+
+  @Override
+  public void perform(@Nonnull final Run run, @Nonnull final FilePath workspace, @Nonnull final Launcher launcher,
+                      @Nonnull final TaskListener listener) throws InterruptedException, IOException
+  {
+    RepositoryManagerV3Client client = null;
+    Map<String, Object> tagAttributes = (isNotBlank(tagAttributesPath) ||
+        isNotBlank(tagAttributesJson)) ? new HashMap<>() : null;
+    EnvVars env = run.getEnvironment(listener);
+
+    try {
+      client = nexus3Client(nexusInstanceId);
+    }
+    catch (RepositoryManagerException e) {
+      e.getResponseMessage();
+      failBuildAndThrow(run, listener, e.getResponseMessage().orElse(e.getMessage()), new IOException(e));
+    }
+
+    if (isNotBlank(tagAttributesPath)) {
+      try {
+        tagAttributes.putAll(parseAttributes(workspace, tagAttributesPath, env));
+      }
+      catch (Exception e) {
+        failBuildAndThrow(run, listener, CreateTag_Error_TagAttributesPath(), new IOException(e));
+      }
+    }
+
+    if (isNotBlank(tagAttributesJson)) {
+      try {
+        tagAttributes.putAll(parseAttributes(tagAttributesJson));
+      }
+      catch (Exception e) {
+        failBuildAndThrow(run, listener, CreateTag_Error_TagAttributesJson(), new IOException(e));
+      }
+    }
+
+    try {
+      Tag tag = client.createTag(tagName, tagAttributes);
+      listener.getLogger().println("Successfully created tag: '" + tag.toJson() + "'");
+    }
+    catch (RepositoryManagerException e) {
+      failBuildAndThrow(run, listener, e.getResponseMessage().orElse(e.getMessage()), new IOException(e));
+    }
+  }
+
+  private Map<String, Object> parseAttributes(final FilePath workspace, final String filePath, final EnvVars env)
+      throws IOException, InterruptedException
+  {
+    return parseAttributes(new FilePath(workspace, env.expand(filePath)).readToString());
+  }
+
+  private Map<String, Object> parseAttributes(String json) {
+    return GSON.fromJson(json, ATTRIBUTE_TYPE);
+  }
+
+  private void failBuild(Run run, TaskListener listener, String reason) {
+    listener.getLogger().println("Failing build due to: " + reason);
+    run.setResult(FAILURE);
+  }
+
+  private void failBuildAndThrow(Run run, TaskListener listener, String reason, IOException ioException)
+      throws IOException
+  {
+    failBuild(run, listener, reason);
+    throw ioException;
+  }
+
+  @Extension
+  @Symbol("createTag")
+  public static final class DescriptorImpl
+      extends BuildStepDescriptor<Builder>
+  {
+    private static final Pattern VALID_TAG_NAME_PATTERN = Pattern.compile("^[0-9a-zA-Z\\-][\\w\\-.]{0,255}$");
+
```

Comment:
Question: @wrightkris you are validating the tag name but I was curious if you also considered validating the attributes size as well as I believe there is a size limit? It sort of starts to fragment off responsibility from the resource in nxrm but I'm internally debating if it should be added for consistency? Just a thought.

Reply:
Name validation was simple to do (via regex); attribute size validation becomes slightly more complex as it's offered via both a file and custom JSON text. I see the thought behind fragmented validation, but it didn't bother me enough to not do it. I think I'd rather drop the name validation than do both, but open to implementing it in the way the majority feel it would be best.